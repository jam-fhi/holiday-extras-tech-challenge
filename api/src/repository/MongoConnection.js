export default class MongoConnection {
	constructor(mongoClient, username, password, host, authdb) {
		this.username = username;
		this.password = password;
		this.host = host;
		this.authDB = authdb;
		this.mongoClient = mongoClient;
	}

	getConnectionOptions() {
		return {
			useNewUrlParser: true,
			numberOfRetries: 10,
			useUnifiedTopology: true
		};
	}

	async getMongoDBConnection() {
		return new Promise((resolve, reject) => {
			this.mongoClient.connect(
				`mongodb://${this.username}:${this.password}@${this.host}/${this.authDB}`,
				this.getConnectionOptions(),
				async (err, client) => {
					if (!err) {
						resolve(client);
					} else {
						reject(err);
					}
				}
			);
		});
	}

	async findOne(dbConn, db, collection, query) {
		const foundDocument = await dbConn
			.db(db)
			.collection(collection)
			.findOne(query);
		return foundDocument;
	}

	async closeConnection(dbConn) {
		await dbConn.close();
	}
}
