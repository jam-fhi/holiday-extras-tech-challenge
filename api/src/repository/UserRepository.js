export default class DatabaseRepository {
	constructor(mongoClient, db, collection) {
		this.mongoClient = mongoClient;
		this.db = db;
		this.collection = collection;
	}

	async getUserByEmailPassword(email, password) {
		const dbConn = await this.mongoClient.getMongoDBConnection();
		const user = await this.mongoClient.findOne(
			dbConn,
			this.db,
			this.collection,
			{
				email,
				password
			}
		);
		await this.mongoClient.closeConnection(dbConn);
		return user;
	}

	async getUserByEmail(email) {
		const dbConn = await this.mongoClient.getMongoDBConnection();
		const user = await this.mongoClient.findOne(
			dbConn,
			this.db,
			this.collection,
			{
				email
			}
		);
		await this.mongoClient.closeConnection(dbConn);
		return user;
	}

	async updateUser(
		_id,
		id,
		email,
		givenName,
		familyName,
		created,
		password,
		about
	) {
		const dbConn = await this.mongoClient.getMongoDBConnection();
		const user = await this.mongoClient.updateOne(
			dbConn,
			this.db,
			this.collection,
			{
				_id
			},
			{
				id,
				email,
				givenName,
				familyName,
				created,
				password,
				about
			}
		);
		await this.mongoClient.closeConnection(dbConn);
		return user;
	}

	async insertUser(id, email, givenName, familyName, created, password, about) {
		const dbConn = await this.mongoClient.getMongoDBConnection();
		const user = await this.mongoClient.insertOne(
			dbConn,
			this.db,
			this.collection,
			{
				id,
				email,
				givenName,
				familyName,
				created,
				password,
				about
			}
		);
		await this.mongoClient.closeConnection(dbConn);
		return user;
	}

	async saveAuthToken(email, password, token) {
		const dbConn = await this.mongoClient.getMongoDBConnection();
		const user = await this.mongoClient.updateOne(
			dbConn,
			this.db,
			this.collection,
			{
				email,
				password
			},
			{
				token
			}
		);
		await this.mongoClient.closeConnection(dbConn);
		return user;
	}
}
