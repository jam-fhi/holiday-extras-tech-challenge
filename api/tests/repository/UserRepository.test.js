import UserRepository from '../../src/repository/UserRepository';
import MongoConnection from '../../src/repository/MongoConnection';
import { dbSetup, dbTeardown } from './fixture/mongoDBFixture';
import {
	validID,
	validGivenName,
	validFamilyName,
	validCreated,
	validAbout,
	validEmail,
	validPwd,
	validToken,
	invalidEmail,
	invalidPwd,
	invalidUnderscoreID,
	validDB,
	validUsername,
	validPassword,
	validHost,
	validAuthDB,
	validNoFoundDocument,
	validInsertDocument,
	validDeleteResult,
	invalidDeleteResult,
	validAllDocsLength
} from '../CommonData';

describe('User Repository', () => {
	const validUpdateByUnderscoreId = true;
	const invalidUpdateByUserscoreId = false;
	const validCollection = 'repoTest';
	let userRepo;

	beforeEach(async () => {
		await dbSetup(
			validUsername,
			validPassword,
			validHost,
			validAuthDB,
			validDB,
			validCollection
		);
		const mongoClient = new MongoConnection(
			validUsername,
			validPassword,
			validHost,
			validAuthDB,
			validDB
		);
		userRepo = new UserRepository(mongoClient, validCollection);
	});

	afterEach(async () => {
		await dbTeardown(
			validUsername,
			validPassword,
			validHost,
			validAuthDB,
			validDB,
			validCollection
		);
	});

	it('Will find one user by email and password', async () => {
		const user = await userRepo.getUserByEmailPassword(validEmail, validPwd);
		delete user._id;
		expect(user).toMatchSnapshot();
	});

	it('Will not find one user by invalid email and password', async () => {
		const user = await userRepo.getUserByEmailPassword(invalidEmail, validPwd);
		expect(user).toBe(validNoFoundDocument);
	});

	it('Will not find one user by email and invalid password', async () => {
		const user = await userRepo.getUserByEmailPassword(validEmail, invalidPwd);
		expect(user).toBe(validNoFoundDocument);
	});

	it('Will find one user by email', async () => {
		const user = await userRepo.getUserByEmail(validEmail);
		delete user._id;
		expect(user).toMatchSnapshot();
	});

	it('Will find not find a user with an invalid email', async () => {
		const user = await userRepo.getUserByEmail(invalidEmail);
		expect(user).toBe(validNoFoundDocument);
	});

	it('Will find one user by _id', async () => {
		const userByEmail = await userRepo.getUserByEmail(validEmail);
		const user = await userRepo.getUserByDBID(userByEmail._id);
		delete user._id;
		expect(user).toMatchSnapshot();
	});

	it('Will not find a user by invalid _id', async () => {
		const user = await userRepo.getUserByDBID(invalidUnderscoreID);
		expect(user).toBe(validNoFoundDocument);
	});

	it('Will update a user', async () => {
		const userByEmail = await userRepo.getUserByEmail(validEmail);
		const userUpdate = await userRepo.updateUser(
			userByEmail._id,
			validID,
			validEmail,
			validGivenName,
			validFamilyName,
			validPassword,
			validAbout
		);
		expect(userUpdate).toBe(validUpdateByUnderscoreId);
	});

	it('Will fail to update a user', async () => {
		const userUpdate = await userRepo.updateUser(
			invalidUnderscoreID,
			validID,
			validEmail,
			validGivenName,
			validFamilyName,
			validPassword,
			validAbout
		);
		expect(userUpdate).toBe(invalidUpdateByUserscoreId);
	});

	it('Will insert a new user', async () => {
		const userInsert = await userRepo.insertUser(
			validID,
			validEmail,
			validGivenName,
			validFamilyName,
			validCreated,
			validPassword,
			validAbout
		);
		expect(userInsert).toBe(validInsertDocument);
	});

	it('Will update the users token', async () => {
		const userUpdate = await userRepo.saveAuthToken(
			validEmail,
			validPwd,
			validToken
		);
		expect(userUpdate).toBe(validInsertDocument);
	});

	it('Will delete a user', async () => {
		const userByEmail = await userRepo.getUserByEmail(validEmail);
		const userDelete = await userRepo.deleteUser(userByEmail._id);
		expect(userDelete).toBe(validDeleteResult);
	});

	it('Will fail to delete a user with an invalid _id', async () => {
		const userDelete = await userRepo.deleteUser(invalidUnderscoreID);
		expect(userDelete).toBe(invalidDeleteResult);
	});

	it('Will find all users', async () => {
		const users = await userRepo.getAllUsers();
		expect(users.length).toBe(validAllDocsLength);
	});
});
