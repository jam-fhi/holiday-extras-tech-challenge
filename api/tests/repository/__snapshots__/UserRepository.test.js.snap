// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Repository Will find one user by email 1`] = `
Object {
  "document": "found",
}
`;

exports[`User Repository Will find one user by email and password 1`] = `
Object {
  "document": "found",
}
`;
